Fix a flaky scheduler preemption e2e

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind flake
/sig scheduling
/cc @ahg-g @liggitt 

**What this PR does / why we need it**:

In the debugging PR (#86325), it turns out in the e2e cluster, sometimes the preemptor Pod can't finish the preemption logic within 1 minute. This PR optimize the logic of checking if the preemption behavior is happening from "check if the Pod is running" to "if the NominatedNodeName field has been set or the Pod is running". Moreover, the preemptor Pod doesn't need to be a ReplicaSet instead of a plain Pod, also the number of low priority Pods has been adjusted to the least necessity manner (it's because from the log, even plain RS Pods need ~45 secs to get them running, so reducing the RS replica number can help). Moreover, increase the timeout from 1 min to 2 mins.

**Which issue(s) this PR fixes**:

Fixes #86068.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

ahg-g reviewed (commented):
thanks.

liggitt commented on test/e2e/scheduling/preemption.go:
```diff
@@ -338,83 +339,78 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
 		// prepare four ReplicaSet
```
nit, three replicasets now

Huang-Wei replied:
done.

liggitt commented on test/e2e/scheduling/preemption.go:
```diff
@@ -338,83 +339,78 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
+			},
+		}
+		preemptorPod := createPod(f, preemptorPodConf)
+		waitForPreemptingWithTimeout(f, preemptorPod, framework.PodGetTimeout)
```
This will return as soon as NominatedNodeName is set on the pod. that happens *before* preempted pods are deleted, which means that this will race with observing the additional rescheduled rs pods in podNamesSeen.

Shouldn't waitForPreemptingWithTimeout wait for `spec.nodeName` to be set instead?

Huang-Wei replied:
I can't think of a good way except for adding a sleeping internal inbetween. As when the podNamesSeen equals to a number for the first time, we can't assert that it's a success - the number may increase which means an unnecessary preemption which is what this case wants to cover.

> Shouldn't waitForPreemptingWithTimeout wait for spec.nodeName to be set instead?

Updated.

liggitt commented on test/e2e/scheduling/preemption.go:
```diff
@@ -338,83 +339,78 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
-		maxRSPodsSeen := []int{5 * 2, 4 * 2, 4}
+		maxRSPodsSeen := []int{1 * 2, 1 * 2, 1}
 		rsPodsSeen := []int{0, 0, 0}
 		for podName := range podNamesSeen {
```
also, this is pre-existing, but while we're here, accessing the podNamesSeen variable here is a data race with the accesses inside the informer handler. That will panic if detected by the go runtime. This shared map between threads needs to coordinate access (e.g. with a [sync.Map](https://golang.org/pkg/sync/#Map))

Huang-Wei replied:
I'm using a buffered channel to resolve this. If extra pods (greater than 6) are signaled, it's also an error.

liggitt commented on test/e2e/scheduling/preemption.go:
```diff
@@ -472,6 +468,33 @@ func createPauseRS(f *framework.Framework, conf pauseRSConfig) *appsv1.ReplicaSe
+}
+
+// waitForPreemptingWithTimeout verifies if 'pod' is preempting within 'timeout', specifically it checks:
+// - if the 'status.NominatedNodeName' field of preemptor 'pod' has been set
```
if this is being used to block before we do the check on the preempted pods, we need to wait for spec.nodeName, see https://github.com/kubernetes/kubernetes/pull/86400/files#r359691588

Huang-Wei replied:
done.

liggitt replied:
this still returns as soon as `len(pod.Status.NominatedNodeName) > 0`, which is too soon. we need to wait for the preempting pod to actually get scheduled (`len(pod.Spec.NodeName) > 0`)

liggitt commented on test/e2e/scheduling/preemption.go:
```diff
@@ -338,83 +339,78 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
 		framework.Logf("pods created so far: %v", podNamesSeen)
 		framework.Logf("length of pods created so far: %v", len(podNamesSeen))
 
 		// count pods number of ReplicaSet{1,2,3}, if it's more than expected replicas
```
even if we wait for spec.nodeName above (meaning sufficient pods were preempted), this still races with the replicationcontroller observing the preempted pod deletions and creating replacements. the `podNamesSeen` needs to tolerate that lag.

Huang-Wei replied:
resolved by the buffered channel.

Huang-Wei commented on test/e2e/framework/replicaset/wait.go:
```diff
@@ -54,6 +55,12 @@ func WaitForReadyReplicaSet(c clientset.Interface, ns, name string) error {
+
+// WaitForReplicaSetTargetAvailableReplicasWithTimeout waits for .status.availableReplicas of a RS to equal targetReplicaNum
+// with given timeout.
+func WaitForReplicaSetTargetAvailableReplicasWithTimeout(c clientset.Interface, replicaSet *appsv1.ReplicaSet, targetReplicaNum int32, timeout time.Duration) error {
```
true...

Huang-Wei commented:
/hold

Huang-Wei commented:
Kindly ping @liggitt

liggitt commented on test/e2e/scheduling/preemption.go:
```diff
@@ -307,8 +308,8 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
-	ginkgo.It("runs ReplicaSets to verify preemption running path [Flaky]", func() {
-		podNamesSeen := make(map[string]struct{})
+	ginkgo.It("runs ReplicaSets to verify preemption running path", func() {
+		podNamesSeen := make(chan string, 6)
```
is there a downside to making this way oversized (e.g. 100)? I don't want to block adds

Huang-Wei replied:
This's an additional gate to block the "under-preempted" case, but a little "brute forcing"... So no downside if we make it oversized.

liggitt commented on test/e2e/scheduling/preemption.go:
```diff
@@ -427,7 +425,7 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
-			if got > expected {
+			if got != expected {
 				framework.Failf("pods of ReplicaSet%d have been over-preempted: expect %v pod names, but got %d", i+1, expected, got)
 			}
```
should this be:
```
if got > expected {
  framework.Failf("pods of ReplicaSet%d have been over-preempted: expect %v pod names, but got %d", i+1, expected, got)
} else if got < expected {
  framework.Failf("pods of ReplicaSet%d have been under-preempted: expect %v pod names, but got %d", i+1, expected, got)
}
```

Huang-Wei replied:
This is better.

Huang-Wei commented:
/retest

liggitt commented:
https://github.com/kubernetes/kubernetes/pull/86400#discussion_r364332248 is still outstanding, lgtm otherwise

Huang-Wei committed:
Fix a flaky scheduler preemption e2e

- Use preemptor pod's Status.NominatedNodeName to signal success of the Preemption behavior
- Optimize the test to eliminate unnecessary Pods creation
- Increase timeout from 1 minute to 2 minutes

Huang-Wei commented:
> [#86400 (comment)](https://github.com/kubernetes/kubernetes/pull/86400#discussion_r364332248) is still outstanding, lgtm otherwise

Resolved. PTAL.

liggitt commented on test/e2e/scheduling/preemption.go:
```diff
@@ -327,94 +328,91 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
 		framework.Logf("length of pods created so far: %v", len(podNamesSeen))
 
+		time.Sleep(10 * time.Second)
+		close(podNamesSeen)
```
why close this? that creates a panic in the AddFunc

Huang-Wei replied:
If not closing it, the `for range podNamesSeen` loop will hang there waiting for the close.

It's like M producers (AddFunc) generating values, and one receiver says "ok, stop sending me more values". One solution is to close both controller and podNamesSeen here: (the `defer close(stopCh)` should also be adjusted):

```go
close(stopCh)
close(podNamesSeen)
```

Or the other option is to put the`for range podNamesSeen` loop into a goroutine:

```go
go podController.Run(stopCh)
defer func() {
	close(stopCh)
	close(podNamesSeen)
}()

...

maxRSPodsSeen := []int{1 * 2, 1 * 2, 1}
rsPodsSeen := []int{0, 0, 0}
go func() {
	for podName := range podNamesSeen {
		...
	}
}()

time.Sleep(10 * time.Second)
for i, got := range rsPodsSeen {
	...
}
```

WDTY?

liggitt replied:
how about:

1. make individual vars to hold counts per rs:
```
var rs1, rs2, rs3 int32
```

2. increment those directly in AddFunc
```
if strings.HasPrefix(podName, "rs-pod1") {
	atomic.AddInt32(&rs1, 1)
} else if strings.HasPrefix(podName, "rs-pod2") {
	atomic.AddInt32(&rs2, 1)
} else if strings.HasPrefix(podName, "rs-pod3") {
	atomic.AddInt32(&rs3, 1)
}
```

3. after doing the preemption and waiting for the preempting pod to be scheduled, wait until the counts match our expectations:

```
err := wait.Poll(time.Second, wait.TestForeverTimeout, func() (bool, error) {
  if rs1 < 2 {
    framework.Logf("waiting for rs1 to observe 2 pod creations, got %d", rs1)
    return false, nil
  } else if rs1 > 2 {
    return false, fmt.Errorf("rs1 had more than 2 pods created: %d", rs1)
  }

  if rs2 < 2 {
    framework.Logf("waiting for rs2 to observe 2 pod creations, got %d", rs2)
    return false, nil
  } else if rs2 > 2 {
    return false, fmt.Errorf("rs2 had more than 2 pods created: %d", rs2)
  }

  if rs3 < 1 {
    framework.Logf("waiting for rs3 to observe 1 pod creations, got %d", rs3)
    return false, nil
  } else if rs3 > 1 {
    return false, fmt.Errorf("rs3 had more than 1 pods created: %d", rs3)
  }
})
if err != nil {
  framework.Failf("failed pod observation expectations: %v", err)
}
```

Huang-Wei replied:
(3) might be problematic: suppose at time T1, all the conditions are satisfied, then we return and think it's a success. However, the chances are: at a later time T2, the backfilled pods (rs1, rs2, rs3) might be over-preempted, i.e. the pods number may exceed the expected ones.

The other way is to check only once after a fixed interval (10 seconds let's say). But either way is not perfect.

liggitt replied:
after 3 succeeds, waiting a fixed period (e.g. 5 seconds), then doing the check again to make sure nothing changed sounds fine

Huang-Wei replied:
SGTM will update. Thanks @liggitt.

Huang-Wei replied:
Updated. PTAL.

liggitt commented on test/e2e/scheduling/preemption.go:
```diff
@@ -472,6 +472,32 @@ func createPauseRS(f *framework.Framework, conf pauseRSConfig) *appsv1.ReplicaSe
+		if err != nil {
+			return false, err
+		}
+		if len(pod.Spec.NodeName) > 0 {
```
hmm... if we want to make sure this actually triggered an eviction, I guess we should wait for both nominatedNodeName and nodeName:

```
if len(pod.Spec.NominatedNodeName) == 0 {
  // log no preemption triggered yet
  return false, nil
}
if len(pod.Spec.NodeName) == 0 {
  // log preemption not complete yet
  return false, nil
}
```

Huang-Wei replied:
But the chances are "pod.Spec.NominatedNodeName" might be set and unset within our checking interval - only "pod.Spec.NodeName" is the final stable state.

liggitt replied:
ah, I didn't realize nominated node name got unset. ok.

Huang-Wei commented:
/retest

liggitt commented on test/e2e/scheduling/preemption.go:
```diff
@@ -327,108 +329,119 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
+		// "*2" means pods of ReplicaSet{1,2} are expected to be only preempted once.
+		expectedRSPods := []int32{1 * 2, 1 * 2, 1}
+		err := wait.Poll(framework.Poll, framework.PollShortTimeout, func() (bool, error) {
+			for i, got := range podNamesSeen {
```
```go
for i, got := range podNamesSeen {
  got := atomic.LoadInt32(&podNamesSeen[i])
```

Huang-Wei replied:
I changed it to:

```go
for i := 0; i < len(podNamesSeen); i++ {
    got := atomic.LoadInt32(&podNamesSeen[i])
```

liggitt commented on test/e2e/scheduling/preemption.go:
```diff
@@ -327,108 +329,119 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
+		// If logic continues to here, we should do a final check to ensure within a time period,
+		// the state is stable; otherwise, pods may be over-preempted.
+		time.Sleep(5 * time.Second)
+		for i, got := range podNamesSeen {
```
```go
for i, got := range podNamesSeen {
  got := atomic.LoadInt32(&podNamesSeen[i])
```

Huang-Wei replied:
I changed it to:

```go
for i := 0; i < len(podNamesSeen); i++ {
    got := atomic.LoadInt32(&podNamesSeen[i])
```

liggitt commented:
/lgtm
/approve
/hold for retesting to flush out any last minute flakes

per https://github.com/kubernetes/kubernetes/issues/86068#issuecomment-566226260 we had a ~10% flake rate before marking this flaky. I'd like to see ~20 test runs of pull-kubernetes-e2e-gce without this test failing before merging this

Huang-Wei commented:
Thanks @liggitt for reviewing.

> per [#86068 (comment)](https://github.com/kubernetes/kubernetes/issues/86068#issuecomment-566226260) we had a ~10% flake rate before marking this flaky. I'd like to see ~20 test runs of pull-kubernetes-e2e-gce without this test failing before merging this

To confirm, I will manually run `/test pull-kubernetes-e2e-gce` 20 times, and ensure the preemption e2e doesn't fail in any of the runnings, correct?

liggitt commented:
> To confirm, I will manually run `/test pull-kubernetes-e2e-gce` 20 times, and ensure the preemption e2e doesn't fail in any of the runnings, correct?

Exactly. sorry for the toil, but I want to make sure the test passes in the exact environment it was having trouble before. (note that I expect *other* currently flaky test failures will be encountered, I just want to make sure this particular test doesn't fail)

Huang-Wei commented:
> Exactly. sorry for the toil, but I want to make sure the test passes in the exact environment it was having trouble before. (note that I expect _other_ currently flaky test failures will be encountered, I just want to make sure this particular test doesn't fail)

Np at all. I will do that.

Huang-Wei commented:
Automatic CI test [0](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/86400/pull-kubernetes-e2e-gce/1216777029093953538/): pass (flaky test is "[sig-apps] DisruptionController should block an eviction until the PDB is updated to allow it")

/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-kind-ipv6

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
/test pull-kubernetes-e2e-gce

Huang-Wei commented:
@liggitt I tried to test the flaky tests for 20 times (https://github.com/kubernetes/kubernetes/pull/86400#issuecomment-573809427) and they all get passed (failures of each are caused by other flakes)

So I'm gonna 
/hold cancel

liggitt commented:
Thanks for the diligence and for the fix
